scene/main/viewport.cpp

```diff
@@ -2183,23 +2196,60 @@ Window *Viewport::get_base_window() const {
 
 	return w;
 }
-void Viewport::_gui_remove_focus_for_window(Node *p_window) {
+void Viewport::_gui_remove_focus_for_window(Node *p_window, int focus_layer) {
 	if (get_base_window() == p_window) {
-		gui_release_focus();
+		gui_release_focus(focus_layer);
+	}
+}
+
+bool Viewport::_gui_control_has_focus(const Control *p_control, int focus_layer) {
+	for (const FocusedControl &E : gui.focused_controls) {
+		if (E.focus_layer == focus_layer) {
+			return E.p_focused_control == p_control;
+		}
+	}
+	return false;
+}
+
+bool Viewport::_gui_has_active_focus_layer() const {
+	for (const FocusedControl &E : gui.focused_controls) {
+		if (E.focus_layer == gui_get_active_focus_layer()) {
+			return true;
+		}
+	}
+	return false;
+}
+
+Control *Viewport::_gui_get_active_focus_control() const {
+	for (const FocusedControl &E : gui.focused_controls) {
+		if (E.focus_layer == gui.active_focus_layer) {
+			return E.p_focused_control;
+		}
 	}
+	return nullptr;
 }
 
-bool Viewport::_gui_control_has_focus(const Control *p_control) {
-	return gui.key_focus == p_control;
+void Viewport::_gui_set_focus_control(Control *p_control, int focus_layer) {
+	for (FocusedControl &E : gui.focused_controls) {
+		if (E.focus_layer == focus_layer) {
+			E.p_focused_control = p_control;
+			return;
+		}
+	}
+	// Add new layer.
+	FocusedControl new_focused_control = FocusedControl();
+	new_focused_control.p_focused_control = p_control;
+	new_focused_control.focus_layer = focus_layer;
+	gui.focused_controls.push_back(new_focused_control);
 }
 
-void Viewport::_gui_control_grab_focus(Control *p_control) {
-	if (gui.key_focus && gui.key_focus == p_control) {
+void Viewport::_gui_control_grab_focus(Control *p_control, int focus_layer) {
```

Comment:
The second parameter here is not necessary, because you can just get the layer from the provided control.